File: src/main/java/ru/stankin/mj/view/AccountWindow.java
package ru.stankin.mj.view;

import com.vaadin.data.fieldgroup.FieldGroup;
import com.vaadin.data.util.BeanItem;
import com.vaadin.data.validator.AbstractStringValidator;
import com.vaadin.ui.*;
import ru.stankin.mj.model.user.User;
import ru.stankin.mj.model.Student;

import java.util.function.BiConsumer;

/**
 * Created by nickl on 16.02.15.
 */
public class AccountWindow extends Window {

    User user;

    BiConsumer<User, String> save;

    public AccountWindow(User user, BiConsumer<User, String> save) {
        this(user, save, false, false);
    }

    public AccountWindow(User user, BiConsumer<User, String> save, boolean warn, boolean needChangePassword ) {
        super("Аккаунт: " + user.getUsername());
        this.user = user;
        this.save = save;
        this.setModal(true);
        this.setDraggable(false);
        this.setResizable(false);
        VerticalLayout vertical = new VerticalLayout();
        vertical.setMargin(true);
        boolean warnPassword = warn && needChangePassword;
        if (warnPassword) {
            vertical.addComponent(new Label("ВНИМАНИЕ!"));
            vertical.addComponent(new Label("На вашем аккаунте используется пароль по умолчанию, пожалуйста, смените его!"));
        }
        final FormLayout content = new FormLayout();
        final FieldGroup binder = new FieldGroup(new BeanItem<>(user));
        content.addComponent(readOnlyField("Логин:", user.getUsername()));
        //content.addComponent(new TextField("Пароль:", new ));
        Field<String> password = new TextField("Пароль:", "");
        //binder.buildAndBind("Пароль:", "password");
        //password.setRequired(true);
        if (warnPassword)
            password.addValidator(new AbstractStringValidator("Неправильный пароль") {
                @Override
                protected boolean isValidValue(String value) {
                    return !value.isEmpty() && !value.equals(user.getUsername());
                }
            });
        content.addComponent(password);
        if (user instanceof Student) {
            Student student = (Student) user;
            content.addComponent(readOnlyField("Группа:", student.stgroup));
            content.addComponent(readOnlyField("Фамилия:", student.surname));
            content.addComponent(readOnlyField("Имя:", student.name));
            content.addComponent(readOnlyField("Отчество:", student.patronym));
            content.addComponent(readOnlyField("Инициалы:", student.initials));
        }
        vertical.addComponent(content);
        vertical.addComponent(
                new HorizontalLayout(
                        new Button("Сохранить", e -> {
                            if (binder.isValid()) {
                                try {
                                    binder.commit();
                                } catch (FieldGroup.CommitException e1) {
                                    e1.printStackTrace();
                                }
                                this.save.accept(user, password.getValue());
                                this.close();
                            }
                        }),
                        new Button("Отменить", event -> this.close()))
        );

        this.setContent(vertical);
        this.center();
    }


    private TextField readOnlyField(String caption, String value) {
        TextField filed = new TextField(caption, value);
        filed.setReadOnly(true);
        return filed;
    }


}

File: src/main/java/ru/stankin/mj/view/AccountWindow.kt
package ru.stankin.mj.view

import com.vaadin.data.Validatable
import com.vaadin.data.fieldgroup.FieldGroup
import com.vaadin.data.util.BeanItem
import com.vaadin.data.validator.AbstractStringValidator
import com.vaadin.server.UserError
import com.vaadin.ui.*
import ru.stankin.mj.model.AuthenticationsStore
import ru.stankin.mj.model.user.User
import ru.stankin.mj.model.Student
import ru.stankin.mj.model.UserResolver

import java.util.function.BiConsumer

/**
 * Created by nickl on 16.02.15.
 */
class AccountWindow @JvmOverloads constructor(
        internal var user: User,
        internal var userDAO: UserResolver,
        internal var auth: AuthenticationsStore,
        adminMode: Boolean = false, requireOldPassword: Boolean = !adminMode) : Window("Аккаунт: " + user.username) {

    init {
        this.isModal = true
        this.isDraggable = false
        this.isResizable = false
        val vertical = VerticalLayout()
        vertical.setMargin(true)
        val warnPassword = !adminMode && auth.acceptPassword(user.id, user.username)
        if (warnPassword) {
            vertical.addComponent(Label("ВНИМАНИЕ!"))
            vertical.addComponent(Label("На вашем аккаунте используется пароль по умолчанию, пожалуйста, смените его!"))
        }
        if(!adminMode && user.email.isNullOrEmpty())
            vertical.addComponent(Label("На вашем аккаунте не указан адрес электроной почты, пожалуйста, укажите адрес, " +
                    "по которому можно с вами связаться!"))

        val content = FormLayout()
        val binder = FieldGroup(BeanItem(user))
        content.addComponent(readOnlyField("Логин:", user.username))
        if (user is Student) {
            val student = user as Student
            content.addComponent(readOnlyField("Группа:", student.stgroup))
            content.addComponent(readOnlyField("Фамилия:", student.surname))
            content.addComponent(readOnlyField("Имя:", student.name))
            content.addComponent(readOnlyField("Отчество:", student.patronym))
            content.addComponent(readOnlyField("Инициалы:", student.initials))
        }
        val email = (binder.buildAndBind("email", "email") as TextField).apply {
            description = "Укажите адрес электронной почты, он пригодится если вы забудете пароль."
        }
        if(user.email == null)
            email.value = ""

        content.addComponent(email)
        //content.addComponent(new TextField("Пароль:", new ));


        val password = (if(adminMode) TextField("Новый пароль:", "") else PasswordField("Новый пароль:", "")).apply {
            description = "Оставьте поле пустым если не хотите менять пароль"
            addValidator(object : AbstractStringValidator("Пароль не должен совпадать с логином") {
                override fun isValidValue(value: String): Boolean {
                    return adminMode || value != user.username
                }
            })
        }

        val removeAuths = CheckBox("Сбросить вход через внешние сервисы")

        val oldPassword = if (requireOldPassword) PasswordField("Текущий Пароль:", "").apply {
            description = "Укажите текущий пароль к вашему аккаунту"
            addValidator(object : AbstractStringValidator("Необходимо указать верный пароль") {
                override fun isValidValue(value: String): Boolean = when {
                    password.value.isEmpty() && !removeAuths.value && !email.isModified-> true
                    else -> auth.acceptPassword(user.id, value)
                }
            })
        } else null


        oldPassword?.let {
            content.addComponent(it)
        }

        content.addComponent(password)
        content.addComponent(removeAuths)
        vertical.addComponent(content)
        vertical.addComponent(
                HorizontalLayout(
                        Button("Сохранить") { e ->
                            oldPassword?.markAsDirty()
                            if (binder.isValid && content.mapNotNull { it as? Validatable }.all { it.isValid }) {
                                try {
                                    binder.commit()
                                } catch (e1: FieldGroup.CommitException) {
                                    e1.printStackTrace()
                                }

                                this.userDAO.saveUserAndPassword(user, password.value)
                                if(removeAuths.value){
                                    auth.dropExternalAuths(user.id)
                                }
                                this.close()
                            }
                        },
                        Button("Отменить") { event -> this.close() })
        )

        this.content = vertical
        this.center()
    }


    private fun readOnlyField(caption: String, value: String): TextField {
        val filed = TextField(caption, value)
        filed.isReadOnly = true
        return filed
    }


}
